### ISSUE TITLE
UNC path support

### ISSUE THREAD

**Issue by Bilge** [composer/composer#9858]
When I run this command: 

```
composer i
```

I get the following output: 

```
'\\wsl$\Ubuntu-18.04\home\bilge'
CMD.EXE was started with the above path as the current directory.
UNC paths are not supported.  Defaulting to Windows directory.
Composer could not find a composer.json file in C:\Windows
To initialize a project, please create a composer.json file as described in the https://getcomposer.org/ "Getting Started" section
```

And I expected this to happen:
UNC path is supported.

---
With the increasing popularity of WSL and with respect to Docker support for WSL 2, it may be increasingly common to want to access the WSL container's files from outside the container. UNC paths are used to address the WSL files from outside the container, such as `\\wsl$\Ubuntu-18.04\home\bilge` in this case.

Of course, we could just use Composer from _inside_ the container, but we may have very good reasons for not doing so, such as having our Composer cache and configuration outside the container, plus some commands such as `browse` will simply never work within the container either.

**Comment by stof** [comment]
Can you share the verbose output ?

**Comment by Bilge** [comment]
@stof
```
$ composer i -vvv
'\\wsl$\Ubuntu-18.04\home\bilge\onejourney-api'
CMD.EXE was started with the above path as the current directory.
UNC paths are not supported.  Defaulting to Windows directory.
Reading C:/Users/Bilge/AppData/Roaming/Composer/composer.json
Loading config file C:/Users/Bilge/AppData/Roaming/Composer/composer.json
Checked CA file C:\Users\Bilge\AppData\Local\Temp\opeCC88.tmp: valid
Executing command (C:/Users/Bilge/AppData/Roaming/Composer): git branch -a --no-color --no-abbrev -v
Executing command (C:/Users/Bilge/AppData/Roaming/Composer): git describe --exact-match --tags
Executing command (CWD): git --version
Executing command (C:/Users/Bilge/AppData/Roaming/Composer): git log --pretty="%H" -n1 HEAD --no-show-signature
Executing command (C:/Users/Bilge/AppData/Roaming/Composer): hg branch
Executing command (C:/Users/Bilge/AppData/Roaming/Composer): fossil branch list
Executing command (C:/Users/Bilge/AppData/Roaming/Composer): fossil tag list
Executing command (C:/Users/Bilge/AppData/Roaming/Composer): svn info --xml
Reading C:/Users/Bilge/AppData/Roaming/Composer/composer.json
Loading config file C:/Users/Bilge/AppData/Roaming/Composer/composer.json
Reading C:/Users/Bilge/AppData/Roaming/Composer/vendor/composer/installed.json
Reading C:/Users/Bilge/AppData/Roaming/Composer/vendor/composer/installed.json
Running 2.0.7 (2020-11-13 17:31:06) with PHP 7.4.2 on Windows NT / 10.0
Reading C:/Users/Bilge/AppData/Roaming/Composer/composer.json
Loading config file C:/Users/Bilge/AppData/Roaming/Composer/composer.json
Executing command (C:/Users/Bilge/AppData/Roaming/Composer): git branch -a --no-color --no-abbrev -v
Executing command (C:/Users/Bilge/AppData/Roaming/Composer): git describe --exact-match --tags
Executing command (C:/Users/Bilge/AppData/Roaming/Composer): git log --pretty="%H" -n1 HEAD --no-show-signature
Executing command (C:/Users/Bilge/AppData/Roaming/Composer): hg branch
Executing command (C:/Users/Bilge/AppData/Roaming/Composer): fossil branch list
Executing command (C:/Users/Bilge/AppData/Roaming/Composer): fossil tag list
Executing command (C:/Users/Bilge/AppData/Roaming/Composer): svn info --xml
Reading C:/Users/Bilge/AppData/Roaming/Composer/composer.json
Loading config file C:/Users/Bilge/AppData/Roaming/Composer/composer.json
Reading C:/Users/Bilge/AppData/Roaming/Composer/vendor/composer/installed.json
Reading C:/Users/Bilge/AppData/Roaming/Composer/vendor/composer/installed.json
Composer could not find a composer.json file in C:\Windows
To initialize a project, please create a composer.json file as described in the https://getcomposer.org/ "Getting Started" section
```

**Comment by stof** [comment]
hmm, this message happens before any log message of composer, so this does not really help finding where it comes from.

A google search made me find https://github.com/AtomLinter/linter-php/issues/221 as a place being affected in a similar way in the past. Unfortunately, the microsoft support link seems to be dead.

**Comment by Seldaek** [comment]
I can reproduce this in PowerShell if running php in windows, composer.phar being in windows filesystem, but CWD being in `\\wsl$\...`.

The warning you're seeing (`CMD.EXE was started with the above path as the current directory. UNC paths are not supported.  Defaulting to Windows directory.`) appears to be output by composer/xdebug-handler but I can not repro anymore with the latest Composer version, so make sure you upgrade and it should be gone. It seems the recent switch to proc_open in xdebug-handler fixed this (cc @johnstevenson).

That said, I noticed the issues with is_readable affecting network mounts (https://bugs.php.net/bug.php?id=73543) on windows seem to also affect WSL paths, so I opened a PR fixing that #9861

**Comment by Bilge** [comment]
This issue was reported using Composer v2.0.7, however the issue is the same on the latest 2.0.13.

**Comment by johnstevenson** [comment]
Could you run it with `XDEBUG_HANDLER_DEBUG=1` env, just so I can see the xdebug-handler output.

**Comment by Bilge** [comment]
```
$ XDEBUG_HANDLER_DEBUG=1 composer -V
xdebug-handler[23976] Checking COMPOSER_ALLOW_XDEBUG
xdebug-handler[23976] The Xdebug extension is loaded (2.9.8)
xdebug-handler[23976] CTRL signals suppressed
xdebug-handler[23976] Process restarting (COMPOSER_ALLOW_XDEBUG=internal|2.9.8|0|*|*)
xdebug-handler[23976] Running "C:\prog\PHP 7.4\php.exe" -n -c C:\Users\Bilge\AppData\Local\Temp\377A.tmp C:/ProgramData/ComposerSetup/bin/composer.phar -V --ansi
'\\wsl$\Ubuntu-18.04\home\bilge'
CMD.EXE was started with the above path as the current directory.
UNC paths are not supported.  Defaulting to Windows directory.
xdebug-handler[19620] Checking COMPOSER_ALLOW_XDEBUG
xdebug-handler[19620] Restarted (217 ms). The Xdebug extension is not loaded
Composer version 2.0.13 2021-04-27 13:11:08
xdebug-handler[23976] Restarted process exited 0
```

**Comment by Seldaek** [comment]
Sorry 2.0.13 still uses Xdebug handler 1.4.6 but if you try to self-update --snapshot you should get a version including xdebug-handler 2.0.0 i believe that is what fixed it for me.

**Comment by johnstevenson** [comment]
@Seldaek Can you tell me how to reproduce this error please.

**Comment by Seldaek** [comment]
As I wrote above

> I can reproduce this in PowerShell if running php in windows, composer.phar being in windows filesystem, but CWD being in \\wsl$\....

Not sure what else to say. Start powershell, cd in wsl path, run composer show where composer is running from windows php/phar and not as wsl binary.

**Comment by johnstevenson** [comment]
Okay now (it didn't like `\\wsl$` on its own - unlike File Explorer).

**Comment by Bilge** [comment]
Indeed, it seems you tend to have to address a specific image in most applications.

**Comment by johnstevenson** [comment]
I'm still working through this, but from what I can see the safest way to resolve the issue is to use `pushd \\wsl$\path` before calling composer then `popd` afterwards. This maps the network path using a temporary drive letter, so that cmd can work in the current directory, rather than defaulting to the Windows directory.

Whether all Composer commands will work is another matter because PHP will be doing Windowsy things in a Linux environment.

cmd can get invoked in several ways:

1. by calling `composer ...` (if Composer is on the path) which uses the composer.bat proxy script (but not if you are using Git Bash)
2. by xdebug-handler if it needs to restart and is not at version 2 running PHP 7.4+ (which calls php directly)
3. when Composer calls any of the process execution functions

1 and 2 are the obvious worst case scenarios, because Composer will start in C:\Windows, but 3 may have unintended consequences because the new process will also run there.

I'll fix xdebug-handler to not restart if we will end up in the Windows directory.

**Comment by Seldaek** [comment]
Ok that sounds good, IMO this is fairly non-critical though as it's not such a mainstream use case so I will ship the fixes in 2.1.0 and not as bugfixes as it depends on xdebug-handler 2..
